core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
@lindong28 : My suggestion is that nobody should directly call LogManager.handleLogDirFailure() except for ReplicaManager.handleLogDirFailure(). The latter knows how to bring all parts to a consistent state with respect to IOException, no matter where it's introduced. So, if a method hits an IOException through ReplicaManager, we will just the IOException bubble up to ReplicaManager and call ReplicaManager.handleLogDirFailure() there. If the IOException is isolated LogCleaner, we can somehow pass ReplicaManager.handleLogDirFailure() to LogCleaner and let LogCleaner call it. This way, any time a disk error is detected, it will be handled consistently no matter in which component the error is detected, which seems easier to reason about.